File: frontend/src/types/invoice.ts
/**
 * Invoice Types for Lisa Roofing CRM
 *
 * Data models for invoices, line items, and payments.
 */

/**
 * Job summary for invoice display (from backend).
 */
export interface JobSummaryForInvoice {
  id: string
  title: string
  jobNumber: string
  customerName: string | null
  customerEmail: string | null
  propertyAddress: string | null
}

/**
 * Invoice status values.
 */
export type InvoiceStatus =
  | 'draft'
  | 'sent'
  | 'viewed'
  | 'partial'
  | 'paid'
  | 'overdue'
  | 'cancelled'

/**
 * Payment method options.
 */
export type PaymentMethod =
  | 'cash'
  | 'check'
  | 'card'
  | 'bank_transfer'
  | 'financing'
  | 'other'

/**
 * Payment status values.
 */
export type PaymentStatus = 'pending' | 'completed' | 'failed' | 'refunded'

/**
 * Line item in an invoice.
 */
export interface InvoiceLineItem {
  id: string
  invoiceId: string
  description: string
  quantity: number
  unitPrice: number
  total: number
  order: number
  createdAt: string
  updatedAt: string
}

/**
 * Payment received for an invoice.
 */
export interface Payment {
  id: string
  invoiceId: string
  amount: number
  paymentDate: string
  paymentMethod: PaymentMethod
  referenceNumber: string | null
  notes: string | null
  status: PaymentStatus
  createdAt: string
  updatedAt: string
}

/**
 * Invoice - Full model with line items and payments.
 */
export interface Invoice {
  id: string
  jobId: string
  invoiceNumber: string
  status: InvoiceStatus
  invoiceDate: string
  dueDate: string | null
  subtotal: number
  taxRate: number
  taxAmount: number
  discountAmount: number
  total: number
  amountPaid: number
  balanceDue: number
  notes: string | null
  terms: string | null
  pdfUrl: string | null
  sentAt: string | null
  paidAt: string | null
  createdAt: string
  updatedAt: string
  // Related data
  lineItems: InvoiceLineItem[]
  payments: Payment[]
  job?: JobSummaryForInvoice
}

/**
 * Invoice summary for list views (lighter payload).
 */
export interface InvoiceSummary {
  id: string
  jobId: string
  invoiceNumber: string
  status: InvoiceStatus
  invoiceDate: string
  dueDate: string | null
  total: number
  amountPaid: number
  balanceDue: number
  customerName: string
  jobTitle: string
}

/**
 * Data for creating a new invoice.
 */
export interface CreateInvoiceData {
  jobId: string
  invoiceDate?: string
  dueDate?: string
  taxRate?: number
  discountAmount?: number
  notes?: string
  terms?: string
}

/**
 * Data for updating an invoice.
 */
export interface UpdateInvoiceData {
  invoiceDate?: string
  dueDate?: string
  taxRate?: number
  discountAmount?: number
  notes?: string
  terms?: string
  status?: InvoiceStatus
}

/**
 * Data for creating a line item.
 */
export interface CreateLineItemData {
  description: string
  quantity: number
  unitPrice: number
  order?: number
}

/**
 * Data for updating a line item.
 */
export interface UpdateLineItemData {
  description?: string
  quantity?: number
  unitPrice?: number
  order?: number
}

/**
 * Data for recording a payment.
 */
export interface CreatePaymentData {
  amount: number
  paymentDate: string
  paymentMethod: PaymentMethod
  referenceNumber?: string
  notes?: string
}

/**
 * Invoice status display configuration.
 */
export const INVOICE_STATUS_CONFIG: Record<
  InvoiceStatus,
  { label: string; color: string; bgColor: string }
> = {
  draft: { label: 'Draft', color: 'text-gray-700', bgColor: 'bg-gray-100' },
  sent: { label: 'Sent', color: 'text-blue-700', bgColor: 'bg-blue-100' },
  viewed: { label: 'Viewed', color: 'text-purple-700', bgColor: 'bg-purple-100' },
  partial: { label: 'Partial', color: 'text-yellow-700', bgColor: 'bg-yellow-100' },
  paid: { label: 'Paid', color: 'text-green-700', bgColor: 'bg-green-100' },
  overdue: { label: 'Overdue', color: 'text-red-700', bgColor: 'bg-red-100' },
  cancelled: { label: 'Cancelled', color: 'text-gray-500', bgColor: 'bg-gray-100' },
}

/**
 * Payment method display labels.
 */
export const PAYMENT_METHOD_LABELS: Record<PaymentMethod, string> = {
  cash: 'Cash',
  check: 'Check',
  card: 'Credit/Debit Card',
  bank_transfer: 'Bank Transfer (ACH)',
  financing: 'Financing',
  other: 'Other',
}

/**
 * Format currency value.
 */
export function formatCurrency(amount: number): string {
  return new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: 'USD',
  }).format(amount)
}

/**
 * Check if invoice is overdue.
 */
export function isInvoiceOverdue(invoice: Invoice | InvoiceSummary): boolean {
  if (invoice.status === 'paid' || invoice.status === 'cancelled') {
    return false
  }
  if (invoice.dueDate) {
    return new Date(invoice.dueDate) < new Date()
  }
  return false
}

File: backend/app/schemas/invoice.py
"""Invoice and Payment schemas."""

from datetime import datetime, date
from uuid import UUID

from pydantic import BaseModel

from app.models.invoice import InvoiceStatus, PaymentMethod


# Payment Schemas
class PaymentCreate(BaseModel):
    """Create payment schema."""
    amount: float
    payment_date: date | None = None
    payment_method: PaymentMethod = PaymentMethod.CARD
    reference_number: str | None = None
    notes: str | None = None


class PaymentResponse(BaseModel):
    """Payment response schema."""
    id: UUID
    invoice_id: UUID
    amount: float
    payment_date: date
    payment_method: PaymentMethod
    reference_number: str | None
    notes: str | None
    stripe_payment_id: str | None
    status: str
    created_at: datetime

    class Config:
        from_attributes = True


# Invoice Line Item Schemas
class InvoiceLineItemCreate(BaseModel):
    """Create invoice line item schema."""
    description: str
    quantity: float = 1.0
    unit_price: float = 0.0
    order: int = 0


class InvoiceLineItemResponse(BaseModel):
    """Invoice line item response schema."""
    id: UUID
    invoice_id: UUID
    description: str
    quantity: float
    unit_price: float
    total: float
    order: int

    class Config:
        from_attributes = True


# Invoice Schemas
class InvoiceCreate(BaseModel):
    """Create invoice schema."""
    job_id: UUID
    invoice_date: date | None = None
    due_date: date | None = None
    tax_rate: float = 0.0
    discount_amount: float = 0.0
    notes: str | None = None
    terms: str | None = None
    line_items: list[InvoiceLineItemCreate] | None = None


class InvoiceUpdate(BaseModel):
    """Update invoice schema."""
    status: InvoiceStatus | None = None
    invoice_date: date | None = None
    due_date: date | None = None
    tax_rate: float | None = None
    discount_amount: float | None = None
    notes: str | None = None
    terms: str | None = None


class JobSummaryForInvoice(BaseModel):
    """Minimal job info for invoice display."""
    id: UUID
    title: str
    job_number: str
    customer_name: str | None = None
    customer_email: str | None = None
    property_address: str | None = None

    class Config:
        from_attributes = True


class InvoiceResponse(BaseModel):
    """Invoice response schema."""
    id: UUID
    job_id: UUID
    invoice_number: str
    status: InvoiceStatus
    
    # Dates
    invoice_date: date
    due_date: date | None
    sent_at: datetime | None
    paid_at: datetime | None
    
    # Amounts
    subtotal: float
    tax_rate: float
    tax_amount: float
    discount_amount: float
    total: float
    amount_paid: float
    balance_due: float
    
    # Content
    notes: str | None
    terms: str | None
    
    # URLs
    pdf_url: str | None
    payment_link: str | None
    
    # Integration
    stripe_invoice_id: str | None
    quickbooks_invoice_id: str | None
    quickbooks_sync_status: str | None
    
    # Related
    line_items: list[InvoiceLineItemResponse]
    payments: list[PaymentResponse]
    job: JobSummaryForInvoice | None = None
    
    is_overdue: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True
